### PR TITLE
Create a shortcut editor for the notebook.

### PR DESCRIPTION
**Pull request by Carreau**
Replace #1265 

Also introduce es6, and write the new keyboard shortcut editor as pure es6.
Does not use react.

--- 

[edit] 
Update description after lots of refactor.

Create shortcut editor for the notebook

1) finish the step allowing the use of es6

  - this include some tweak to web pack configuration to speed up
      recompile in watch mode (in particular cache sourcemaps).
  - enable eslint (error only), on obvious mistakes.
  - setup babel to compile to es5 as a target.

2) Make the test pass under Casper that does not always have
`Function.prototype.bind` defined, which we cannot patch only in the
tests.

3) Write an actual shortcut editor that list and allow to modify most of
the command mode shortcut.

The logic to persist the shortcuts is a bit tricky as there are default
keyboard shortcuts, and so when you "unbind" them you need to re-unbind
them at next startup. This does not work for a few shortcut for
technical reasons: `<Esc>`, `<Shift>`, as well as `<Ctrl-Shift-P>` and `<F>`
which register asynchronously, so are not detected as "default"
shortcuts.

### CODE REVIEW

**Comment by sccolbert** [comment]
Should preact be a runtime dependency instead of a dev dependency?

**Comment by Carreau** [comment]
> Should preact be a runtime dependency instead of a dev dependency?

Likely yes, reflect of `--save-dev`.

**Comment by Carreau** [comment]
reflex*,

and it will be bundled in the min.js that will generated by webpack and be shipped, so not actually not sure.

File: package.json [hunk 1]
```diff
@@ -27,5 +30,7 @@
   },
   "dependencies": {
     "moment": "^2.8.4"
```
Comment (sccolbert) [inline package.json]: this is a syntax error, you need a comma

File: package.json [hunk 2]
```diff
@@ -27,5 +30,7 @@
...
+    "preact": "^4.5.1",
+    "preact-compat": "^1.7.0",
```
Comment (sccolbert) [inline package.json]: No trailing comma.

**Comment by Carreau** [comment]
ping @jdfreder , you're the one that know the most things about notebook test. Test seem to not run at all with this PR, but I can't seem to figure out the reason why, as the same step works flawlessly locally.

Any ideas ?

**Comment by Carreau** [comment]
BTW, this now works, and can now use es6.

**Comment by sccolbert** [comment]
What do you mean by "can use es6"?

**Comment by Carreau** [comment]
> What do you mean by "can use es6"?

Sorry, was unclear, you can use es6 feature in all the js files (typically `notebook/static/*/js/*.js`), webpack use the a babel loader when minifying, so will understand `let`, `const`, `import {} from ...` etc. There is no extra steps.

**Comment by sccolbert** [comment]
Ah - I didn't realize we were using babel. Are we still targeting ES5 runtimes?

**Comment by Carreau** [comment]
> Are we still targeting ES5 runtimes?

Yes. We can use something else than babel, I guess we can use a typescript loader as well, I have no issues with that. 

One thing I did not include is the ability to use `.jsx`, but that should be pretty easy to add too if we decide to do so.

**Comment by sccolbert** [comment]
I don't think we need to use the TS loader at this point. We've been doing ago just shipping the built `.js` files for our code which is written in TS.

**Comment by sccolbert** [comment]
Also, I think (but could be wrong) that preact has an `h()` function, which can reduce the boilerplate needed by `createElement`.

**Comment by Carreau** [comment]
> Also, I think (but could be wrong) that preact has an h() function, which can reduce the boilerplate needed by createElement.

Yes it has a `h()` function, though they don't really explain how/why to use it, expect using `/** @jsx h */` in `jsx` files, and it looks mostly as `s/createElement/h/g`, but I might be missing something.

**Comment by Carreau** [comment]
Ok, merging to get people to complain.

**Pull request merged**